### ISSUE TITLE
`comparision-overlap` ignores side-effects from member function calls

### ISSUE THREAD

**Issue by hterik** [python/mypy#19783]
**To Reproduce**
`mypy --strict` on file below gives following false error:
```
scratch_43.py:20:8: error: Non-overlapping equality check (left operand type: "Literal[NodeStatus.Offline]", right operand type: "Literal[NodeStatus.Running]")  [comparison-overlap]
```

```python
import dataclasses
from enum import Enum

class NodeStatus(Enum):
    Offline = 1
    Running = 2
    Available = 3

@dataclasses.dataclass
class Node:
    status: NodeStatus = NodeStatus.Offline

    def set_free(self) -> None:
        self.status = NodeStatus.Available

def process_node(node: Node) -> None:
    if node.status != NodeStatus.Offline:
        return                                  # This narrows node.status to Literal[NodeStatus.Offline] forever

    node.set_free()                             # <<< mypy ignores side effects on .status from this
                                                #     This should restore node.status to any Nodestatus
    if node.status != NodeStatus.Running:       # error:  Non-overlapping equality check (left operand type: "Literal[NodeStatus.Offline]", right operand type: "Literal[NodeStatus.Running]")  [comparison-overlap]
        node.status = NodeStatus.Running
```

**Expected Behavior**

Reading `node.status` after calling `node.set_free()` should reset previous narrowing of `node.status`.

**Actual Behavior**

Narrowing done early inside a function can not be reset through mutating functions.

**Your Environment**

- Mypy version used: `1.17`, `1.14`, `1.13`
- Mypy command-line flags: `--strict`
- Python version used: 1.13
